### PR TITLE
Add support for skipped columns

### PR DESCRIPTION
**Pull request by matiasfrndz**
While working with the ageing WIP chart I observed unexpected results when issues in Jira had skipped columns. Looking deeper into the issue I noticed that this also resulted in unexpected results in the calculation of cycle times. To demonstrate an correct the unexpected results I added two new tests:

- test_calculate_ageing_wip_with_skipped_columns in ageingwip_test
- test_movement_skipped_columns in cycletime_test

The change at hand corrects the results of both tests.

This is an updated pull request based on #24 and the feedback provided @rafalotufo in #23.

### CODE REVIEW

**Comment by optilude** [comment]
Both - I’m really grateful for your contributions here. I’m struggling for time but keeping an eye. If you can both test and confirm it is working (and automated tests are passing) I’ll merge again.

**Comment by matiasfrndz** [comment]
Thanks @optilude , I'll check with @rafalotufo and let you know.

@rafalotufo, you mentioned the burnup chart being broken by this change. Can you elaborate, who I can reproduce the error(s)?

**Comment by matiasfrndz** [comment]
I think I found the problem @rafalotufo. I had made unnecessary changes to `burnup.py` and `forecast.py` which i reverted in 7c12081. Could you check if it works for you? Thank you in advance!

File: jira_agile_metrics/calculators/cycletime_test.py [hunk 1]
```diff
@@ -88,6 +88,27 @@ def jira(custom_fields):
...
+            status=Value("Done", "done"),
+            resolution=Value("Done", "Done"),
+            resolutiondate="2018-01-04 01:01:01",
+            created="2018-01-01 01:01:01",
+            customfield_001="Team 1",
+            customfield_002=Value(None, 10),
+            customfield_003=Value(None, []),
+            customfield_100=None,
+            changes=[
+                Change("2018-01-02 01:05:01", [("status", "Backlog", "Next",)]),
```
Comment (rafalotufo) [inline jira_agile_metrics/calculators/cycletime_test.py:106]: This is the test I think that needs to be changed to reproduce the issue I'm talking about. If you change the transition from Backlog to Build, I think the test will fail, while the age of the ticket hasn't really changed (it's started work on 2018-01-02, though the first stage was Build and not Committed).

File: jira_agile_metrics/calculators/cycletime_test.py [hunk 2]
```diff
@@ -107,6 +107,22 @@ def jira_with_skipped_columns(custom_fields):
...
+            created="2018-01-01 01:01:01",
+            customfield_001="Team 1",
+            customfield_002=Value(None, 10),
+            customfield_003=Value(None, []),
+            customfield_100=None,
+            changes=[
+                Change("2018-01-02 01:05:01", [("status", "Backlog", "Build",)]),
+                Change("2018-01-04 01:01:01", [("status", "Build", "Done",), ("resolution", None, "done")]), # skipping columns Build and Test
+            ],
+        ),
```
Comment (matiasfrndz) [inline jira_agile_metrics/calculators/cycletime_test.py:125]: I added the case you described @rafalotufo . All tests pass without changes to any other code.

**Comment by matiasfrndz** [comment]
@rafalotufo ,  I added the adapted test as described. Does it work as expected now? Thank you in advance for your support. 👍🏼

**Comment by rafalotufo** [comment]
Hi Matias,

If you make this change:

*diff --git a/jira_agile_metrics/conftest.py
b/jira_agile_metrics/conftest.py*

*index cbdebff..5caa6c6 100644*

*--- a/jira_agile_metrics/conftest.py*

*+++ b/jira_agile_metrics/conftest.py*

@@ -256,7 +256,7 @@ def
large_cycle_time_results(minimal_cycle_time_columns):

             dict(Backlog=_ts('2018-01-01'), Committed=_ts('2018-01-02'),
Build=_ts('2018-01-04'), Test=NaT,               Done=NaT),



             # three issues in test

-            dict(Backlog=_ts('2018-01-01'), Committed=_ts('2018-01-02'),
Build=_ts('2018-01-03'), Test=_ts('2018-01-04'), Done=NaT),

+            dict(Backlog=_ts('2018-01-01'), Committed=NaT,
Build=_ts('2018-01-02'), Test=_ts('2018-01-04'), Done=NaT),

             dict(Backlog=_ts('2018-01-01'), Committed=_ts('2018-01-02'),
Build=_ts('2018-01-03'), Test=_ts('2018-01-05'), Done=NaT),

             dict(Backlog=_ts('2018-01-01'), Committed=_ts('2018-01-02'),
Build=_ts('2018-01-03'), Test=_ts('2018-01-05'), Done=NaT),


You will see that 3 tests fail. The first one that fails is
test_calculate_ageing_wip since it will not contain the issue that had the
skipped Committed stage.

Let me know if you have any other questions. I recomment the fix that I
had, which simply orders the dates and uses the first one instead of
relying on a "committed" state.

Rafael

**Comment by optilude** [comment]
I know we’re just discussing here but can we please remember to add new tests rather than change existing ones, so that we don’t accidentally cause other regressions?

---
Martin Aspeli
Partner | Deloitte Digital
+44 07989 304475
________________________________

**Comment by optilude** [comment]
Also: the distinction between initial state and committed is important functionally. The two can be the same, but often won’t be (and if you are running a proper kanban process, shouldn’t be!). It’s important that we always treat “committed” as the start point for cycle time calculations and other metrics, not “backlog”.

Therefore I’m also not quite clear in my head what is supposed happen if someone skips the committed stage. It’s possible we need to treat it differently but it requires more thinking...

---
Martin Aspeli
Partner | Deloitte Digital
+44 07989 304475
________________________________

**Comment by matiasfrndz** [comment]
@optilude, thanks for the feedback. That's exactly what I thought too, thus I added tests covering the situations mentioned by @rafalotufo. I also agree to the comment regarding the commitment point.

**Comment by rafalotufo** [comment]
I think committed can be any state that is different then "Backlog". It's
hard to make assumptions of other people's workflows and this does make a
very strong assumption that there is one specific state that is the
committed one. That is not true for my case. A ticket can go from Backlog
to Committed or from Backlog to Test directly.

**Comment by optilude** [comment]
So to get a bit philosophical here for a moment: in the sense of measuring cycle time in a meaningful way (and thus to be able to do something useful with the charts and data you get from it, such as forecasting or drawing conclusions about team performance), you need two things:

- A defined commitment point, ie when you start the clock ticking. That cannot be the moment of creation in the backlog because by definition the backlog contains things you haven’t prioritised or committed to yet. The committed stage should usually be a queue, ie it means you have prioritised it but not started work yet, though it could be an active stage too in some very efficient pull systems.

- A defined completion point, ie when do you stop the clock.

In this tool you can configure which two JIRA states (or sets of states mapped to a workflow stage, to be precise) describe those two points. If you don’t configure it explicitly, it uses the second and final stages, respectively. This is also why we validate that your workflow has at least three stages.

JIRA of course lets you do all kind of crazy things. If your workflow allows you to jump from any state to any state, or in some other way skip the commitment point, then I don’t think the data you get out of this tool or any other tool is going to be very meaningful.

If what you really want to model is that we committed to it and started work immediately and the first thing we did was test, then the answer is probably to have two transitions and just do them immediately after one another. The calculations always use a resolution of days anyway so the time prior to test in this case would be 0 days.

This may not matter, but I’m mindful we don’t want to create complexity in the code or accidentally create regressions or fail to throw errors for a configuration that isn’t logical.

**Comment by matiasfrndz** [comment]
Thank you for sharing your thoughts on this @optilude , I totally agree with what you've written.

**Comment by matiasfrndz** [comment]
@rafalotufo thank for the suggestion in https://github.com/DeloitteDigitalUK/jira-agile-metrics/pull/26#issuecomment-618112115. I have refactored `cycletime.py` so that it calculates the correct timestamp for skipped columns, mimicking the behaviour of ActionableAgile Analytics. This makes calculations in subsequent calculators much easier, because they do no't have to handle these special cases anymore. Thus, with the proposed change to `cycletime.py` this case doesn't occur anymore:
```python
dict(Backlog=_ts('2018-01-01'), Committed=NaT, Build=_ts('2018-01-02'), Test=_ts('2018-01-04'), Done=NaT),
````
Thank you very much for your feedback and support.

**Comment by matiasfrndz** [comment]
After having received so much valuable feedback, I have decided to close this pull request, refactor the changes based on your feedback, and then create a new and clean pull request.

🙏🏼👍🏼

**Comment by optilude** [comment]
Cool, thanks!

If we are refactoring cycletime.py please take extra care and do what you can to test both in unit tests and in the real world.

Pretty much all of the calculations stem from that algorithm and it has evolved several quirks over time to deal with weirdness in JIRA etc. so is probably a bit more fragile. (I have projects than run off this software so I just get a bit cautious).

Doesn’t mean we shouldn’t fix bugs or make it better of course.

Really appreciate all your work on this!

---
Martin Aspeli
Partner | Deloitte Digital
+44 07989 304475
________________________________

**Pull request closed without merging**